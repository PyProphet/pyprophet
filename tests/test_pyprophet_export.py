from __future__ import print_function

import os
import subprocess
import shutil
import sys

import pandas as pd
import sqlite3

import pytest

pd.options.display.expand_frame_repr = False
pd.options.display.precision = 4
pd.options.display.max_columns = None

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")

def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += str(subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT))
    except subprocess.CalledProcessError as error:
        print(error, end="", file=sys.stderr)
        raise
    return stdout


def _run_osw(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # peptide-level
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " peptide --pi0_lambda=0.001 0 0 --in=test_data.osw --context=global"

    # protein-level
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=run-specific"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=experiment-wide"
    cmdline += " protein --pi0_lambda=0 0 0 --in=test_data.osw --context=global"

    # export
    cmdline += " export --in=test_data.osw --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    if not peptide:
        cmdline += " --no-peptide"

    if not protein:
        cmdline += " --no-protein"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_osw_unscored(regtest, temp_folder, transition_quantification=False, peptide=False, protein=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)

    # export
    cmdline = "pyprophet export --in=test_data.osw --out=test_data.tsv --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_ipf(regtest, temp_folder, transition_quantification=False, ipf="disable"):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)
    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms1 --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # MS2-level
    cmdline += " score --in=test_data.osw --level=ms2 --test --pi0_lambda=0.001 0 0 --ss_iteration_fdr=0.02"

    # transition-level
    cmdline += " score --in=test_data.osw --level=transition --test --pi0_lambda=0.1 0 0 --ss_iteration_fdr=0.02"

    # IPF
    cmdline += " ipf --in=test_data.osw"

    # export
    cmdline += " export --in=test_data.osw --no-peptide --no-protein --ipf_max_peptidoform_pep=1  --max_rs_peakgroup_qvalue=1 --format=legacy_merged"

    if not transition_quantification:
        cmdline += " --no-transition_quantification"

    cmdline += " --ipf=" + ipf

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_unscored(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # export
    cmdline = "pyprophet export-compound --in=test_data_compound.osw --out=test_data_compound_unscored.tsv --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_unscored.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)

def _run_compound_ms1(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS1-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms1 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms1.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms1.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    
def _run_compound_ms2(regtest, temp_folder):
    os.chdir(temp_folder)
    data_path= os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path, temp_folder)
    
    # MS2-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms2 --test"

    # export
    cmdline += " export-compound --in=test_data_compound.osw --out=test_data_compound_ms2.tsv --max_rs_peakgroup_qvalue=0.05 --format=legacy_merged"

    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_compound_ms2.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)


def test_osw_0(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, False)

def test_osw_1(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, True, False, False)

def test_osw_2(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, True, False)

def test_osw_3(tmpdir, regtest):
    _run_osw(regtest, tmpdir.strpath, False, False, True)

def test_ipf_0(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "disable")

def test_ipf_1(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, True, "disable")

def test_ipf_2(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "peptidoform")

def test_ipf_3(tmpdir, regtest):
    _run_ipf(regtest, tmpdir.strpath, False, "augmented")

def test_compound_0(tmpdir, regtest):
    _run_compound_ms1(regtest, tmpdir.strpath)
    
def test_compound_1(tmpdir, regtest):
    _run_compound_ms2(regtest, tmpdir.strpath)

def test_compound_unscored(tmpdir, regtest):
	_run_compound_unscored(regtest, tmpdir.strpath)
	
def test_osw_unscored(tmpdir, regtest):
	_run_osw_unscored(regtest, tmpdir.strpath)
