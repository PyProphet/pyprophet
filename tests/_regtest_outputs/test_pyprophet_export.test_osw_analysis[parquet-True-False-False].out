    Charge             FullPeptideName  Intensity                                        ProteinName       RT        Sequence                           aggr_Fragment_Annotation                                     aggr_Peak_Apex                                     aggr_Peak_Area  aggr_prec_Peak_Apex  aggr_prec_Peak_Area  assay_iRT   assay_rt  d_score  decoy  delta_iRT  delta_rt                         filename      iRT                   id  leftWidth  m_score        mz  peak_group_rank  rightWidth               run_id  transition_group_id
0        2  ADSTGTLVITDPTR(UniMod:267)   207283.0  AQUA4SWATH_HMLangeA;AQUA4SWATH_HMLangeA;AQUA4S...  2661.55  ADSTGTLVITDPTR  1343_b4_1;1359_y3_1;1344_b5_1;1360_y4_1;1369_y...  969.0;36907.0;1426.0;6131.0;416.0;2071.0;11984...  10322.0;251772.0;9915.0;43365.0;5402.0;15040.0...          117220.7500             854645.0       26.5  2595.5788   5.7301  False     1.9379   65.9712  napedro_L120420_010_SW.mzXML.gz  28.4379 -4409520928686189639  2640.5100   0.0033  728.8795              1.0   2705.3701 -8670811102654834151                    0
1        2  ADSTGTLVITDPTR(UniMod:267)     6385.0  AQUA4SWATH_HMLangeA;AQUA4SWATH_HMLangeA;AQUA4S...  2605.74  ADSTGTLVITDPTR  1363_y6_2;1348_b8_2;1343_b4_1;1359_y3_1;1365_y...  0.0;520.0;181.0;1338.0;294.0;20.0;232.0;10.0;1...  150.0;13418.0;3123.0;9325.0;2329.0;100.0;4328....            8790.7810             104006.0       26.5  2595.5733   1.2404  False     0.3198   10.1667  napedro_L120420_010_SW.mzXML.gz  26.8198   260819276075322832  2575.6399   0.0685  728.8795              2.0   2623.4399 -8670811102654834151                    0
2        2  ADSTGTLVITDPTR(UniMod:267)     3838.0  AQUA4SWATH_HMLangeA;AQUA4SWATH_HMLangeA;AQUA4S...  2708.53  ADSTGTLVITDPTR  1363_y6_2;1348_b8_2;1343_b4_1;1359_y3_1;1365_y...  0.0;541.0;294.0;1060.0;90.0;0.0;190.0;30.0;429...  50.0;5415.0;1984.0;6992.0;1344.0;60.0;1748.0;1...            5750.4717              73215.2       26.5  2595.5750   0.8151  False     3.3002  112.9550  napedro_L120420_010_SW.mzXML.gz  29.8002  8534214264242363560  2705.3701   0.2018  728.8795              3.0   2736.0901 -8670811102654834151                    0
3        2  ADSTGTLVITDPTR(UniMod:267)     2693.0  AQUA4SWATH_HMLangeA;AQUA4SWATH_HMLangeA;AQUA4S...  2795.06  ADSTGTLVITDPTR  1363_y6_2;1348_b8_2;1343_b4_1;1359_y3_1;1365_y...  0.0;702.0;334.0;386.0;241.0;10.0;212.0;20.0;29...  10.0;4454.0;1675.0;2411.0;1323.0;60.0;1121.0;1...            4036.5600              25862.3       26.5  2595.5754   0.6777  False     5.8092  199.4846  napedro_L120420_010_SW.mzXML.gz  32.3092  6932937885234622359  2790.7200   0.2018  728.8795              4.0   2811.2000 -8670811102654834151                    0
4        2  ADSTGTLVITDPTR(UniMod:267)     5180.0  AQUA4SWATH_HMLangeA;AQUA4SWATH_HMLangeA;AQUA4S...  2832.77  ADSTGTLVITDPTR  1363_y6_2;1348_b8_2;1343_b4_1;1359_y3_1;1365_y...  0.0;592.0;397.0;743.0;427.0;20.0;222.0;0.0;348...  30.0;10869.0;5111.0;8116.0;33319.0;271.0;6193....           10419.7430             241873.0       26.5  2595.5778  -0.1013  False     6.9026  237.1922  napedro_L120420_010_SW.mzXML.gz  33.4026  5163914660633416481  2811.2000   0.4692  728.8795              5.0   2855.5801 -8670811102654834151                    0
..     ...                         ...        ...                                                ...      ...             ...                                                ...                                                ...                                                ...                  ...                  ...        ...        ...      ...    ...        ...       ...                              ...      ...                  ...        ...      ...       ...              ...         ...                  ...                  ...
95       2      VYVYAVDQTR(UniMod:267)   597887.0  AQUA4SWATH_HMLangeA;AQUA4SWATH_HMLangeA;AQUA4S...  2230.18      VYVYAVDQTR  1817_y6_2;1808_b7_2;1819_y7_2;1810_b8_2;1812_b...  439.0;40.0;968.0;120.0;765.0;916.0;20367.0;193...  3589.0;571.0;4040.0;965.0;6525.0;4613.0;89094....          269150.5600            1192530.0       16.3  2245.2318   5.6532  False    -0.3700  -15.0518  napedro_L120420_010_SW.mzXML.gz  15.9300  6870255268859409918  2213.2600   0.0033  612.3184              1.0   2247.3999 -8670811102654834151                   19
96       2      VYVYAVDQTR(UniMod:267)    16553.0  AQUA4SWATH_HMLangeA;AQUA4SWATH_HMLangeA;AQUA4S...  2317.38      VYVYAVDQTR  1817_y6_2;1803_b3_1;1808_b7_2;1813_y3_1;1819_y...  224.0;847.0;10.0;582.0;30.0;50.0;295.0;531.0;5...  2071.0;5859.0;141.0;4805.0;321.0;503.0;2775.0;...            9656.7600              89588.0       16.3  2245.2345   0.3846  False     2.1584   72.1455  napedro_L120420_010_SW.mzXML.gz  18.4584  6262215160571261022  2302.0200   0.2507  612.3184              2.0   2332.7400 -8670811102654834151                   19
97       2      VYVYAVDQTR(UniMod:267)    20746.0  AQUA4SWATH_HMLangeA;AQUA4SWATH_HMLangeA;AQUA4S...  2120.97      VYVYAVDQTR  1817_y6_2;1803_b3_1;1808_b7_2;1813_y3_1;1819_y...  283.0;913.0;50.0;1597.0;60.0;50.0;643.0;937.0;...  4611.0;12157.0;330.0;11363.0;594.0;1098.0;5460...           10568.0810              85676.6       16.3  2245.2373   0.3343  False    -3.5368 -124.2673  napedro_L120420_010_SW.mzXML.gz  12.7632  -610141049182829192  2100.6001   0.4000  612.3184              3.0   2141.5701 -8670811102654834151                   19
98       2      VYVYAVDQTR(UniMod:267)    48058.0  AQUA4SWATH_HMLangeA;AQUA4SWATH_HMLangeA;AQUA4S...  2291.53      VYVYAVDQTR  1817_y6_2;1803_b3_1;1808_b7_2;1813_y3_1;1819_y...  397.0;263.0;50.0;520.0;162.0;80.0;441.0;296.0;...  4032.0;2661.0;580.0;6989.0;1362.0;732.0;6854.0...            6480.8610              75465.1       16.3  2245.2345  -0.0155  False     1.4088   46.2955  napedro_L120420_010_SW.mzXML.gz  17.7088  2043199813358518344  2267.8799   0.4692  612.3184              4.0   2298.6101 -8670811102654834151                   19
99       2      VYVYAVDQTR(UniMod:267)    10959.0  AQUA4SWATH_HMLangeA;AQUA4SWATH_HMLangeA;AQUA4S...  2538.15      VYVYAVDQTR  1817_y6_2;1803_b3_1;1808_b7_2;1813_y3_1;1819_y...  306.0;233.0;30.0;510.0;40.0;213.0;406.0;562.0;...  1334.0;1410.0;140.0;2288.0;140.0;1194.0;1952.0...           10810.2700              56553.9       16.3  2245.2307  -1.8782  False     8.5599  292.9193  napedro_L120420_010_SW.mzXML.gz  24.8599 -5430403952310232561  2520.5100   0.4692  612.3184              5.0   2544.4099 -8670811102654834151                   19

[100 rows x 27 columns]
